System/FilePath/Internal.hs

```diff
@@ -290,13 +290,24 @@ getSearchPath = fmap splitSearchPath (getEnv "PATH")
 -- > splitExtension "file.txt/boris.ext" == ("file.txt/boris",".ext")
 -- > splitExtension "file/path.txt.bob.fred" == ("file/path.txt.bob",".fred")
 -- > splitExtension "file/path.txt/" == ("file/path.txt/","")
+
+-- A naive implementation would be to use @splitFileName_@ first,
+-- then break filename into basename and extension, then recombine dir and basename.
+-- This is way too expensive, see @splitFileName_@ comment for discussion.
+--
+-- Instead we speculatively split on the extension separator first, then check
+-- whether results are well-formed.
 splitExtension :: FILEPATH -> (STRING, STRING)
-splitExtension x = if null nameDot
-                   then (x, mempty)
-                   else (dir <> init nameDot, singleton extSeparator <> ext)
-    where
-        (dir,file)    = splitFileName_ x
-        (nameDot,ext) = breakEnd isExtSeparator file
+splitExtension x
+  -- Imagine x = "no-dots", then nameDot = ""
+  | null nameDot = (x, mempty)
+  -- Imagine x = "\\shared.with.dots\no-dots"
+  | isWindows && null (dropDrive nameDot) = (x, mempty)
```

Comment:
I know realize that current filepath has this behavior:

```
ghci> W.splitExtension "\\shared.with.dots"
("\\shared.with",".dots")
```

This is wrong and a bug. But I don't know if we can fix this without breaking code.

Reply:
https://github.com/haskell/filepath/issues/185

Reply:
And I found another bug:

```
ghci> W.splitExtension "\\.\foo"
("\\",".\foo")
```

`\\.\` on windows is special like `\\?\`.

Reply:
Oh forget it... I forgot that ghci needs escaping of slashes, lol.